basesandbox/src/ipc/multiplex.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright 2020 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use super::{IpcRecv, IpcSend, RecvError, Terminate};
+use crossbeam::channel::bounded;
+use std::thread;
+
+type Sender = crossbeam::channel::Sender<Vec<u8>>;
+type Receiver = crossbeam::channel::Receiver<Vec<u8>>;
+
+pub trait Forward {
+    fn forward(data: &[u8]) -> usize;
+}
+
+fn sender<S: IpcSend>(send: S, recv: Receiver) {
+    loop {
+        let data = recv.recv().unwrap();
+        // special exit flag
+        if data.is_empty() {
+            break
+        }
+        send.send(&data);
+    }
+}
+
+fn receiver<F: Forward, R: IpcRecv>(recv: R, send: Vec<Sender>) {
+    loop {
+        let data = match recv.recv(None) {
+            Err(RecvError::TimeOut) => panic!(),
+            Err(RecvError::Termination) => return,
+            Ok(x) => x,
+        };
+        send[F::forward(&data)].send(data).unwrap();
+    }
+}
+
+pub struct Multiplexer {
+    sender_thread: Option<thread::JoinHandle<()>>,
+    receiver_thread: Option<thread::JoinHandle<()>>,
+    termiantor: Option<Box<dyn Terminate>>,
+    sender_send: Sender, // will convey the flag of termination
+}
+
+impl Multiplexer {
+    pub fn create<F: Forward, S: IpcSend + 'static, R: IpcRecv + 'static>(
+        ipc_send: S,
+        ipc_recv: R,
+        multiplex: usize,
+        channel_capacity: usize,
+    ) -> (Vec<(Sender, Receiver)>, Self) {
+        let (sender_send, sender_recv) = bounded(channel_capacity);
+        let mut receiver_sends = Vec::<Sender>::new();
+        let mut channel_ends = Vec::<(Sender, Receiver)>::new();
```

Comment:
How about introduce a ChannelEnd struct?
It is hard to read what is the role of Sender and Receiver.

Reply:
Well even I if I do so the field name will be just `sender` and `receiver`

Reply:
OK